src/v/cloud_storage/inventory/report_parser.cc

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cloud_storage/inventory/report_parser.h"
+
+#include "base/vassert.h"
+
+#include <seastar/util/short_streams.hh>
+
+#include <zlib.h>
+
+namespace {
+constexpr auto newline{'\n'};
+} // namespace
+
+namespace cloud_storage::inventory {
+
+report_parser::report_parser(
+  ss::input_stream<char> stream,
+  gzip_compressed_t is_compressed,
+  size_t chunk_size)
+  : _stream{std::move(stream)}
+  , _is_compressed{is_compressed}
+  , _chunk_size{chunk_size} {}
+
+ss::future<std::optional<ss::sstring>> report_parser::next_line() {
+    while (!_buffer.contains(newline)) {
+        co_await fill_buffer();
+        if (eof()) {
+            break;
+        }
+    }
+
+    if (_buffer.empty()) {
+        co_return std::nullopt;
+    }
+
+    if (eof() && !_buffer.contains(newline)) {
+        ss::sstring last_row;
+        _buffer.swap(last_row);
+        co_return last_row;
+    }
+
+    vassert(
+      _buffer.contains(newline),
+      "Buffer {} does not contain a newline",
+      _buffer);
+
+    const auto nl_idx = _buffer.find(newline);
+    auto head = _buffer.substr(0, nl_idx);
+
+    _buffer = _buffer.substr(nl_idx + 1);
+
+    co_return head;
+}
+
+ss::future<> report_parser::fill_buffer() {
+    if (_is_compressed) {
+        if (unlikely(!_decompression_started)) {
+            initialize_decompression();
+        }
+        co_await decompress_chunk();
+    } else {
+        auto buf = co_await _stream.read_up_to(_chunk_size);
+        if (buf.empty()) {
+            _eof = true;
+            co_return;
+        }
+        _buffer.append(buf.get(), buf.size());
+    }
+}
+
+void report_parser::initialize_decompression() {
+    vassert(!_decompression_started, "decompression initialized twice");
+    constexpr auto default_windowbits = 15;
+    constexpr auto decode_with_header_detection = 32;
+    if (auto init_res = inflateInit2(
+          &_zs, default_windowbits + decode_with_header_detection);
+        init_res != Z_OK) {
+        throw report_decompression_error{fmt::format(
+          "Failed to initialize decompression context: {}", zError(init_res))};
+    }
+
+    _decompression_started = true;
+}
+
+report_parser::~report_parser() {
+    if (_decompression_started) {
+        inflateEnd(&_zs);
+    }
+}
+
+ss::future<> report_parser::decompress_chunk() {
```

Comment:
> Just curious if you've looked through /compression/internal/gzip_compressor.cc if anything can be reused from there?

When I started on this PR I assumed that the existing gzip compressor could be used here but it turns out that we can't use it directly. That code is built around decompressing a single iobuf at a time, so it creates one z_stream per iobuf, assumes all the data to be inflated is in that iobuf, inflates it and closes the stream.

In our case the inventory file is almost certainly not going to fit in memory in one iobuf, we are going to receive it as a stream from the `remote::download_...` method. So what we do here is initialize the z_stream for the entire input stream up front, then read and inflate chunks from it on demand, as the consumer of the report parser asks for lines. 

The z_stream is destroyed only after the full report is consumed and must keep it's internal state around while the report is being consumed.

The code in this PR is actually based on the existing gzip compressor, with the changes to handle a stream instead of a single iobuf. I thought about keeping it in the compression lib but assumed this is a niche use case. 

However if it sounds like decompressing larger streams is useful I can move this utility their. It will require some changes though as the stream decompression code will need to be split out from the report parser.

Reply:
I moved the gzip stream decompression utility to the compression namespce/lib. I did not move the report parser entirely as it is too specific to inventory reports, I don't see parsing CSV reports as something that belongs to the compression namespace.